packages/@aws-cdk/aws-ec2/lib/user-data.ts

```diff
@@ -276,3 +276,258 @@ class CustomUserData extends UserData {
+
+  /**
+   * Constructs the new `MultipartUserDataPart` wrapping existing `UserData`. Modification to `UserData` are reflected
+   * in subsequent renders of the part.
+   *
+   * For more information about content types see `MultipartUserDataPartOptionsWithBody`
+   */
+  public static fromUserData(userData: UserData, opts?: MultipartUserDataPartWrapperOptions): MultipartUserDataPart {
+    opts = opts || {};
+    return new MultipartUserDataPartWrapper(userData, opts);
+  }
+
+  /**
+   * Constructs the raw `MultipartUserDataPart` using specified body, content type and transfer encoding.
+   *
+   * When transfer encoding is specified (typically as Base64), it's caller responsibility to convert body to
+   * Base64 either by wrapping with `Fn.base64` or by converting it by other converters.
+   */
+  public static fromRawBody(opts: MultipartUserDataPartOptionsWithBody): MultipartUserDataPart {
+    return new MultipartUserDataPartRaw(opts);
+  }
+
+  protected static readonly DEFAULT_CONTENT_TYPE = 'text/x-shellscript; charset="utf-8"';
+
+  /** The body of this MIME part. */
+  public abstract get body(): string | undefined;
+
+  /** `Content-Type` header of this part */
+  public readonly contentType: string;
+
+  /**
+   * `Content-Transfer-Encoding` header specifying part encoding.
+   *
+   * @default undefined - don't add this header
+   */
+  public readonly transferEncoding?: string;
+
+  public constructor(props: MultipartUserDataPartOptions) {
+    this.contentType = props.contentType;
+    this.transferEncoding = props.transferEncoding;
+  }
+}
+
+/**
+ * The raw part of multi-part user data, which can be added to {@link MultipartUserData}.
+ */
+class MultipartUserDataPartRaw extends MultipartUserDataPart {
+  private _body : string | undefined;
+
+  public constructor(props: MultipartUserDataPartOptionsWithBody) {
+    super(props);
+    this._body = props.body;
+  }
+
+  public get body(): string | undefined {
+    return this._body;
+  }
+}
+
+/**
+ * Wrapper for `UserData`.
+ */
+class MultipartUserDataPartWrapper extends MultipartUserDataPart {
+  public constructor(public readonly userData: UserData, opts: MultipartUserDataPartWrapperOptions) {
+    super({
+      contentType: opts.contentType || MultipartUserDataPart.DEFAULT_CONTENT_TYPE,
+      // Force Base64 in case userData will contain UTF-8 characters
+      transferEncoding: 'base64',
+    });
+  }
+
+  public get body(): string {
+    // Wrap rendered user data with Base64 function, in case data contains tokens
+    return Fn.base64(this.userData.render());
+  }
+}
+
+/**
+ * Options for creating {@link MultipartUserData}
+ */
+export interface MultipartUserDataOptions {
+  /**
+   * The string used to separate parts in multipart user data archive (it's like MIME boundary).
+   *
+   * This string should contain [a-zA-Z0-9] characters only, and should not be present in any part, or in text content of archive.
+   */
+  readonly partsSeparator: string;
```

Comment:
Can we make this optional and autogenerate it?

Reply:
It's a good question, as the separator should be unique in the whole MIME archive.

_Originally, I tired to generate SHA256 from all parts, but if parts contains tokens, than SHA form such a string is not stable (token id can change), and will trigger change separator, change of separator will trigger change of user data, and change of user data will trigger updates to EC2 or ASG._

---

However I think I've got two ideas

1. Return token, token on resolve will generate random value and store it in context (cons: single instance of separator for project).
2. Return token, which on resolve will resolve parts, and than create hash by hashing the representation of resolution. However because result of _resolve_ can be an object, the stable mechanism for hashing / representing objects  has to be added (i.e. when traversing object graph properties has to inspected in ascending order), etc. Maybe there's such a method in CDK or dependent library.

Simplifying, for following part `'start' + myOtherConstruct.ref + '; next line'`, we could try to SHA following text
`'start' + 'Fn:ref' + 'myOtherId' + ';next line');

As I have hashing part  in commit history, I can try to work on 2, if it's viable.

Reply:
I used here hard coded default separator.